File: src/MyServiceBus.Serialization.Bson/BsonMessageDeserializer.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Bson;

namespace MyServiceBus.Serialization.Bson;

public sealed class BsonMessageDeserializer : IMessageDeserializer
{
    private readonly IMessageHeaderConvention _headerConvention;
    private readonly JsonSerializerSettings _deserializerSettings;

    public BsonMessageDeserializer()
        : this(BsonSerializationSettings.Create(), MassTransitHeaderConvention.Instance)
    {
    }

    public BsonMessageDeserializer(
        JsonSerializerSettings deserializerSettings,
        IMessageHeaderConvention? headerConvention = null)
    {
        ArgumentNullException.ThrowIfNull(deserializerSettings);
        _deserializerSettings = deserializerSettings;
        _headerConvention = headerConvention ?? MassTransitHeaderConvention.Instance;
    }

    public string ContentType => BsonSerializerFactory.BsonContentType;

    public IInboundMessage Deserialize(MessageBody body, IDictionary<string, object> headers)
    {
        ArgumentNullException.ThrowIfNull(body);
        ArgumentNullException.ThrowIfNull(headers);

        try
        {
            using var stream = body.GetStream();
            using var reader = new BsonDataReader(stream);
            var serializer = Newtonsoft.Json.JsonSerializer.Create(_deserializerSettings);
            var envelope = serializer.Deserialize<BsonEnvelope>(reader)
                ?? throw new InvalidDataException("The MassTransit BSON envelope was not found.");
            return new BsonInboundMessage(envelope, headers, _deserializerSettings, _headerConvention);
        }
        catch (BsonSerializationException)
        {
            throw;
        }
        catch (Exception exception)
        {
            throw new BsonSerializationException(
                "Failed to deserialize the MassTransit BSON envelope.",
                exception);
        }
    }

    public MessageBody GetMessageBody(string text)
    {
        try
        {
            return new ByteArrayMessageBody(Convert.FromBase64String(text));
        }
        catch (Exception exception)
        {
            throw new BsonSerializationException("The BSON message body is not valid Base64.", exception);
        }
    }
}

File: src/Java/myservicebus-serialization-bson/src/main/java/com/myservicebus/serialization/bson/BsonMessageDeserializer.java
package com.myservicebus.serialization.bson;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.myservicebus.serialization.ByteArrayMessageBody;
import com.myservicebus.serialization.InboundMessage;
import com.myservicebus.serialization.MessageBody;
import com.myservicebus.serialization.MessageDeserializer;
import com.myservicebus.serialization.MessageHeaderConvention;
import de.undercouch.bson4jackson.BsonFactory;
import java.io.IOException;
import java.util.Base64;
import java.util.Map;
import java.util.UUID;

public final class BsonMessageDeserializer implements MessageDeserializer {
    private final ObjectMapper applicationMapper;
    private final ObjectMapper bsonMapper;
    private final MessageHeaderConvention headerConvention;

    public BsonMessageDeserializer(ObjectMapper applicationMapper, MessageHeaderConvention headerConvention) {
        if (applicationMapper == null) {
            throw new IllegalArgumentException("applicationMapper must not be null");
        }
        if (headerConvention == null) {
            throw new IllegalArgumentException("headerConvention must not be null");
        }
        SimpleModule dotNetBsonModule = new SimpleModule();
        dotNetBsonModule.addDeserializer(UUID.class, new DotNetGuidDeserializer());
        this.applicationMapper = applicationMapper.copy().registerModule(dotNetBsonModule);
        this.bsonMapper = new ObjectMapper(new BsonFactory());
        this.headerConvention = headerConvention;
    }

    @Override
    public String getContentType() {
        return BsonSerializerFactory.BSON_CONTENT_TYPE;
    }

    @Override
    public InboundMessage deserialize(MessageBody body, Map<String, Object> headers) throws IOException {
        try {
            ObjectNode envelope = (ObjectNode) bsonMapper.readTree(body.getBytes());
            if (envelope == null) {
                throw new IOException("The MassTransit BSON envelope was not found.");
            }
            return new BsonInboundMessage(envelope, headers, applicationMapper, headerConvention);
        } catch (Exception exception) {
            if (exception instanceof BsonSerializationException bsonException) {
                throw bsonException;
            }
            throw new BsonSerializationException(
                    "Failed to deserialize the MassTransit BSON envelope.",
                    exception);
        }
    }

    @Override
    public MessageBody getMessageBody(String text) {
        try {
            return new ByteArrayMessageBody(Base64.getDecoder().decode(text));
        } catch (IllegalArgumentException exception) {
            throw new BsonSerializationException("The BSON message body is not valid Base64.", exception);
        }
    }
}
